OCPBUGS#50642: NTP server configuration for HCP (#91222)
openshift/openshift-docs: modules/hcp-configure-ntp.adoc

====
+
.Example `99-worker-chrony.bu` configuration
[source,yaml,subs="attributes+"]
----
variant: openshift
version: {product-version}.0
metadata:
  name: 99-worker-chrony
  labels:
    machineconfiguration.openshift.io/role: worker
storage:
  files:
  - path: /etc/chrony.conf
    mode: 0644 <1>
    overwrite: true
    contents:
      inline: |
        pool 0.rhel.pool.ntp.org iburst <2>
        driftfile /var/lib/chrony/drift
        makestep 1.0 3
        rtcsync
        logdir /var/log/chrony
----
<1> Specify an octal value mode for the `mode` field in the machine config file. After creating the file and applying the changes, the `mode` field is converted to a decimal value.
<2> Specify any valid, reachable time source, such as the one provided by your DHCP server.
+
[NOTE]
====
For all-machine to all-machine communication, the NTP on UDP port is `123`. If you configured an external NTP time server, you must open UDP port `123`.
====

. Use Butane to generate a `MachineConfig` object file, `99-worker-chrony.yaml`, that contains a configuration to be sent to the nodes. Run the following command:
+
[source,terminal]
----
$ butane 99-worker-chrony.bu -o 99-worker-chrony.yaml
----

> 

Reviewer: What do you think about adding an example of the contents generated in 99-worker-chrony.yaml here? That way the user can match it to where to paste this in the ConfigMap below

Author: @CrystalChun SGTM. So this[1] content should be pasted in the config map below? The whole configuration or a specific part from it?
[1] https://github.com/openshift/openshift-docs/pull/91222/files#diff-9f99f3a5d9087ba89c8067c1952964d156c344decca9c846ff5a4f44e86624e2R19

Reviewer: Cool, the linked content you mentioned is actually the butane file 
The contents of the 99-worker-chrony.yaml file should look like this
```yaml
# Generated by Butane; do not edit
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: <machineconfig_name>
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      files:
        - contents:
            compression: ""
            source: data:...
          mode: 420
          overwrite: true
          path: /example/path
```

Reviewer: However, if you think it's redundant to what's in the ConfigMap below then feel free to omit this :) Just thought it might be helpful for the user to know they have to paste the whole yaml content into the `data.config` field of the ConfigMap

Author: I like the idea of adding `.yaml` configuration here. Added ✔️

Reviewer: Awesome, thanks! Also I'm sorry but the example I gave you had a small mistake 😅, added a comment to fix it

[id="hcp-configure-ntp_{context}"]
= Configuring the NTP server for hosted clusters

You can configure the Network Time Protocol (NTP) server for your hosted clusters by using Butane.

.Procedure

. Create a Butane config file, `99-worker-chrony.bu`, that includes the contents of the `chrony.conf` file. For more information about Butane, see "Creating machine configs with Butane".
+
.Example `99-worker-chrony.bu` configuration
[source,yaml,subs="attributes+"]
----
variant: openshift
version: {product-version}.0
metadata:
  name: 99-worker-chrony
  labels:
    machineconfiguration.openshift.io/role: worker
storage:
  files:
  - path: /etc/chrony.conf
    mode: 0644  #<1>
    overwrite: true
    contents:
      inline: |
        pool 0.rhel.pool.ntp.org iburst  #<2>
        driftfile /var/lib/chrony/drift
        makestep 1.0 3
        rtcsync
        logdir /var/log/chrony
----
<1> Specify an octal value mode for the `mode` field in the machine config file. After creating the file and applying the changes, the `mode` field is converted to a decimal value.
<2> Specify any valid, reachable time source, such as the one provided by your DHCP server.
+
[NOTE]
====

> For all-machine to all-machine communication, the NTP on UDP port is `123`. If you configured an external NTP time server, you must open UDP port `123`.

Reviewer 2: ```suggestion
For machine-to-machine communication, the NTP on the User Datagram Protocol (UDP) port is `123`. If you configured an external NTP time server, you must open UDP port `123`.
```

Unless "all-machine-to-all-machine"

====

> . Use Butane to generate a `MachineConfig` object file, `99-worker-chrony.yaml`, that contains a configuration to be sent to the nodes. Run the following command:

Reviewer 2: ```suggestion
. Use Butane to generate a `MachineConfig` object file, `99-worker-chrony.yaml`, that contains a configuration that Butane sends to the nodes. Run the following command:
```

$ butane 99-worker-chrony.bu -o 99-worker-chrony.yaml
----
+
.Example `99-worker-chrony.yaml` configuration
[source,yaml]
----
# Generated by Butane; do not edit
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: <machineconfig_name>
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      files:
        - contents:
            source: data:...
          mode: 420
          overwrite: true
          path: /example/path
----

. Add the contents of the `99-worker-chrony.yaml` file inside of a config map in the management cluster:
+
.Example config map
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: <configmap_name>
  namespace: <namespace>  #<1>
data:
  config: |
    apiVersion: machineconfiguration.openshift.io/v1
    kind: MachineConfig
    metadata:
      labels:
        machineconfiguration.openshift.io/role: worker
      name: <machineconfig_name>
    spec:
      config:
        ignition:
          version: 3.2.0
        storage:
          files:
          - contents:
              source: data:...
            mode: 420
            overwrite: true

>             path: /example/path

Reviewer 2: Complete definition? If not, consider adding:

# ...

At the end of the file.

----

> <1> Replace `<namespace>` with the name of your namespace where you created the node pool, for example, `clusters`.

Reviewer 2: ```suggestion
<1> Replace `<namespace>` with the name of your namespace where you created the node pool, such as `clusters`.
```

. Apply the config map to your node pool by running the following command:
+
[source,terminal]
----
$ oc edit nodepool <nodepool_name> --namespace <hosted_cluster_namespace>
----
+
.Example `NodePool` configuration
[source,yaml]
----
apiVersion: hypershift.openshift.io/v1alpha1
kind: NodePool
metadata:
# ...
  name: nodepool-1
  namespace: clusters
# ...
spec:
  config:
  - name: <configmap_name> #<1>
# ...
----
<1> Replace `<configmap_name>` with the name of your config map.

> . Add the list of your NTP servers in the `infra-env.yaml` file, that defines the `InfraEnv` custom resource (CR):

Reviewer 2: ```suggestion
. Add the list of your NTP servers in the `infra-env.yaml` file, which defines the `InfraEnv` custom resource (CR):
```

Reviewer 2: ![Screenshot From 2025-04-08 12-04-55](https://github.com/user-attachments/assets/aba0ff4d-161d-4e29-9ef4-723835769d89)

+
.Example `infra-env.yaml` file
[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
# ...
spec:
  additionalNTPSources:
  - <ntp_server> #<1>
  - <ntp_server1>
  - <ntp_server2>
# ...
----
<1> Replace `<ntp_server>` with the name of your NTP server. For more details about creating a host inventory and the `InfraEnv` CR, see "Creating a host inventory".

. Apply the `InfraEnv` CR by running the following command:
+
[source,terminal]
----
$ oc apply -f infra-env.yaml
----

.Verification

> * Verify that your host inventory is created by running the following command. Check the following fields in the command output:

Reviewer 2: I do not see a command here but instead descriptions of output parameters.

Reviewer 2: Remove "Verify that your host inventory is created by running the following command"?

Author: The command is placed after parameter descriptions.  `$ oc describe infraenv <infraenv_resource_name> -n <infraenv_namespace>`  
I will shift `Verify that your....by running the...` above the command instead of mentioning it in the beginning.

// Module included in the following assemblies:
// * hosted_control_planes/hcp-machine-config.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-configure-ntp_{context}"]
= Configuring the NTP server for hosted clusters

You can configure the Network Time Protocol (NTP) server for your hosted clusters by using Butane.

.Procedure

. Create a Butane config file, `99-worker-chrony.bu`, that includes the contents of the `chrony.conf` file.
+
[NOTE]
====
See "Creating machine configs with Butane" for information about Butane.
====
+
.Example `99-worker-chrony.bu` configuration
[source,yaml,subs="attributes+"]
----
variant: openshift
version: {product-version}.0
metadata:
  name: 99-worker-chrony
  labels:

>     machineconfiguration.openshift.io/role: worker <1>

Reviewer: ```suggestion
    machineconfiguration.openshift.io/role: worker
```

      inline: |
        pool 0.rhel.pool.ntp.org iburst <2>
        driftfile /var/lib/chrony/drift
        makestep 1.0 3
        rtcsync
        logdir /var/log/chrony
----
<1> Specify an octal value mode for the `mode` field in the machine config file. After creating the file and applying the changes, the `mode` field is converted to a decimal value. You can check the updated field in the YAML file with the `oc get mc <machineconfig_name> -o yaml` command.
<2> Specify any valid, reachable time source, such as the one provided by your DHCP server.
+
[NOTE]
====
For all-machine to all-machine communication, the NTP on UDP port is `123`. If you configured an external NTP time server, you must open UDP port `123`.
====

. Use Butane to generate a `MachineConfig` object file, `99-worker-chrony.yaml`, that contains a configuration to be sent to the nodes. Run the following command:
+
[source,terminal]
----
$ butane 99-worker-chrony.bu -o 99-worker-chrony.yaml
----

. Add the contents of the `99-worker-chrony.yaml` file inside of a config map in the management cluster:
+
.Example config map
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: <configmap_name>

>   namespace: clusters

Reviewer: I'd probably add a note that this is the same namespace as the nodepool

data:
  config: |
    apiVersion: machineconfiguration.openshift.io/v1
    kind: MachineConfig
    metadata:
      labels:
        machineconfiguration.openshift.io/role: worker
      name: <machineconfig_name>
    spec:
      config:
        ignition:
          version: 3.2.0
        storage:
          files:
          - contents:
              source: data:...
            mode: 420
            overwrite: true

>             path: <path> <1>

Reviewer: I think this is already generated and the user needs to literally copy the Machine Config from `99-worker-chrony.yaml` file into this configmap

<5> Replace `<hosted_cluster_namespace>` with the name of your hosted cluster namespace.
<6> Replace `<bundle>` with the name of your trust bundle.
<7> Replace `<os_image_version>` with the version of your Discovery Image.

.. Apply the `InfraEnv` CR by running the following command:
+
[source,terminal]
----
$ oc apply -f infra-env.yaml
----

.Verification

* Verify that your host inventory is created by running the following command. Check the following fields in the command output:
+
** `conditions`:  The standard Kubernetes conditions indicating if the image was created successfully.
** `isoDownloadURL`: The URL to download the Discovery Image.
** `createdTime`: The time at which the image was last created. If you modify the `InfraEnv` CR, ensure that you have updated the timestamp before downloading a new image.
+
[source,terminal]
----
$ oc describe infraenv <infraenv_resource_name> -n <infraenv_namespace>
----
+
[NOTE]
====
If you modify the `InfraEnv` CR, confirm that the `InfraEnv` CR has created a new Discovery Image by looking at the `createdTime` field. If you already booted hosts, boot them again with the latest Discovery Image.

> You can continue by configuring static networking, if required, and begin adding hosts to your infrastructure environment.

Reviewer: I don't think this document needs the static networking line
